lib/spack/spack/parser.py

```diff
@@ -0,0 +1,458 @@
+    COMPILER = auto()
+    # Package name
+    UNQUALIFIED_PACKAGE_NAME = auto()
+    FULLY_QUALIFIED_PACKAGE_NAME = auto()
+    # White spaces
+    WS = auto()
+    # Unexpected character
+    UNEXPECTED = auto()
+
+    def __str__(self):
+        return f"{self._name_}"
+
+
+class Token:
+    """Represents tokens; generated from input by lexer and fed to parse()."""
+
+    __slots__ = "kind", "value", "start", "end"
+
+    def __init__(
+        self, kind: TokenKind, value: str, start: Optional[int] = None, end: Optional[int] = None
+    ):
+        self.kind = kind
+        self.value = value
+        self.start = start
+        self.end = end
+
+    def __repr__(self):
+        return str(self)
+
+    def __str__(self):
+        return f"({self.kind}, {self.value})"
+
+    def __eq__(self, other):
+        return (self.kind == other.kind) and (self.value == other.value)
+
+
+#: List of all the regexes used to match spec parts, in order of precedence
+TOKEN_REGEXES = [
+    # Dependency
+    rf"(?P<{TokenKind.DEPENDENCY}>\^)",
+    # Version regexes
+    rf"(?P<{TokenKind.VERSION_HASH_PAIR}>@({VALID_GIT_VERSION}={VALID_VERSION}))",
+    rf"(?P<{TokenKind.VERSION}>@({VALID_VERSION_LIST}))",
+    # Variant regexes
+    rf"(?P<{TokenKind.PROPAGATED_BOOL_VARIANT}>(\+\+|~~|--){VALID_NAME})",
+    rf"(?P<{TokenKind.BOOL_VARIANT}>[~+-]{VALID_NAME})",
+    rf"(?P<{TokenKind.PROPAGATED_KEY_VALUE_PAIR}>({VALID_NAME}==({VALID_VALUE}|{VALID_QUOTED_VALUE})))",  # noqa: E501
+    rf"(?P<{TokenKind.KEY_VALUE_PAIR}>({VALID_NAME}=({VALID_VALUE}|{VALID_QUOTED_VALUE})))",
+    # Compiler regexes
+    rf"(?P<{TokenKind.COMPILER_AND_VERSION}>%({VALID_NAME})([\s]*)@({VALID_VERSION_LIST}))",
+    rf"(?P<{TokenKind.COMPILER}>%({VALID_NAME}))",
+    # Filename
+    rf"(?P<{TokenKind.FILENAME}>({VALID_FILENAME}))",
+    # Spec name regexes
+    rf"(?P<{TokenKind.FULLY_QUALIFIED_PACKAGE_NAME}>{VALID_NAME_WITH_DOTS})",
+    rf"(?P<{TokenKind.UNQUALIFIED_PACKAGE_NAME}>{VALID_NAME_WITHOUT_DOTS})",
+    # DAG hash
+    rf"(?P<{TokenKind.DAG_HASH}>/({VALID_HASH}))",
+    # White spaces (the lowest priority)
+    rf"(?P<{TokenKind.WS}>\s+)",
+]
+
+ERROR_HANDLING_REGEXES = TOKEN_REGEXES + [
+    rf"(?P<{TokenKind.UNEXPECTED}>(.[\s]*))",
+]
+
+#: Maps a string representation to the corresponding token kind
+STR_TO_TOKEN = {str(x): x for x in TokenKind}
+
+MASTER_REGEX = re.compile("|".join(TOKEN_REGEXES))
+ANALYSIS_REGEX = re.compile("|".join(ERROR_HANDLING_REGEXES))
+
+
+def tokenize(text: str) -> Iterator[Token]:
+    """Return a token generator from the text passed as input.
+
+    Raises:
+        SpecTokenizationError: if we can't tokenize anymore, but didn't reach the
+            end of the input text.
+    """
+    scanner = MASTER_REGEX.scanner(text)  # type: ignore[attr-defined]
+    match: Optional[re.Match] = None
+    for match in iter(scanner.match, None):
+        yield Token(
+            STR_TO_TOKEN[match.lastgroup],  # type: ignore[attr-defined]
```

Comment:
You can avoid `STR_TO_TOKEN` entirely:

```suggestion
            TokenKind.__members__[match.lastgroup],  # type: ignore[attr-defined]
```

Reply:
Done in 0b94c97